pkg/controllermanager/controller/shootstate/add.go

```diff
@@ -0,0 +1,117 @@
+// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package shootstate
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	predicateutils "github.com/gardener/gardener/pkg/controllerutils/predicate"
+)
+
+// ControllerName is the name of the controller.
+const ControllerName = "shootstate-finalizer"
+
+// AddToManager adds the Reconciler to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+
+	return builder.ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(
+			&gardencorev1beta1.ShootState{},
+			builder.WithPredicates(predicateutils.ForEventTypes(predicateutils.Create, predicateutils.Update)),
+		).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: ptr.Deref(r.Config.ConcurrentSyncs, 0),
+		}).
+		Watches(
+			&gardencorev1beta1.Shoot{},
+			handler.EnqueueRequestsFromMapFunc(r.MapShootToShootState),
+			builder.WithPredicates(r.ShootPredicates()),
+		).Complete(r)
+}
+
+// MapShootToShootState maps a Shoot object to ShootState reconciliation request.
+func (r *Reconciler) MapShootToShootState(_ context.Context, obj client.Object) []reconcile.Request {
+	shoot, ok := obj.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return nil
+	}
+
+	namespacedName := types.NamespacedName{
+		Name:      shoot.Name,
+		Namespace: shoot.Namespace,
+	}
+	return []reconcile.Request{{NamespacedName: namespacedName}}
+}
+
+// ShootPredicates returns predicates for Shoot requests acceptance.
+func (r *Reconciler) ShootPredicates() predicate.Predicate {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
```

Comment:
If you do not define `CreateFunc`, `DeleteFunc` and `GenericFunc` they will be defaulted to true and this will enqueue the mapper function.

Reply:
Thanks for pointing it, added the missing predicates in my latest updates.